### PR TITLE
Convert StatusBarRegistration to use components field

### LINKED ISSUE #1571: Extension API statusBarItems doesn't register status bar items
**Describe the bug**

I have an extension using `statusBarItems` api <https://docs.k8slens.dev/master/extensions/capabilities/common-capabilities/#status-bar-items>. but no icons are shown on the status bar after I launch Lens.

Can't see the status bar item in DOM tree as well, so I guess is not a CSS issue but related to extension manage/loader?

**To Reproduce**
1. Open Lens
2. No status bar item

**Expected behavior**

An icon on status bar.

**Screenshots**

Screenshot 2020-11-30 at 16 45 54

**Environment (please complete the following information):**
- Lens Version: 4.0.0-rc.1
- OS: Big Sur
- Installation method (e.g. snap or AppImage in Linux): .dmg
 
**Logs:**

No error logs in developer tools console.

**Additional context**

The source code of my extension
```tsx
import { LensRendererExtension, Interface, Component } from "@k8slens/extensions";
import React from "react"

const { Icon } = Component;

const StatusBarItemIcon = ({ navigate }: { navigate?: () => void }): JSX.Element => (
  <Icon
    material="link"
    interactive
    style={{
      color: "rgb(255, 255, 0)"
    }}
    onClick={() => navigate()}
  />
)

export default class RendererExtension extends LensRendererExtension {
  statusBarItems: Interface.StatusBarRegistration[] = [
    {
      item: (): JSX.Element => <StatusBarItemIcon navigate={this.navigate.bind(this)} />,
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by Nokel81**
- More similar to all other *Registration types for extensions

- Simpler fix for using the components.Icon type, now accepts functions
  that return component instance like all other *Registration types

- Fixed not linting the whole repo, including fixing a type decl

Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #1571

### CODE REVIEW

File: .eslintrc.js
Comment (jim-docker) [inline .eslintrc.js]: is this required to fulfill this PR's title? Otherwise could be done in a separate PR where it doesn't muddy things?

File: src/extensions/registries/status-bar-registry.ts
Comment (jim-docker) [inline src/extensions/registries/status-bar-registry.ts]: Why `Icon`? Why not `Item`?

Comment (Nokel81) [reply src/extensions/registries/status-bar-registry.ts]: I guess I was caught up with the specific example I was working with. Good catch.

File: .eslintrc.js
Comment (Nokel81) [reply .eslintrc.js]: Sounds good.

**Comment by jim-docker** [comment]
Doesn't look like it impacts aqua's extension but it does impact @stefcameron's lens-extension-cc

File: docs/extensions/capabilities/common-capabilities.md
Comment (jim-docker) [inline docs/extensions/capabilities/common-capabilities.md]: `Item`

File: docs/extensions/guides/renderer-extension.md
Comment (jim-docker) [inline docs/extensions/guides/renderer-extension.md]: `Item`

**Review by jim-docker — COMMENTED** [review]
LGTM but @jakolehm should decide when this is merged

File: src/extensions/registries/status-bar-registry.ts
Comment (stefcameron) [inline src/extensions/registries/status-bar-registry.ts:11]: It feels strange to have this name be plural. Even based on the type def, we're still talking about a _singular_ component. Whether that "component" is a fragment that's technically an array of components (but still, I believe, seen as a single component within the realms of React), or is a classical or functional single component, it's still just one component, right...?

**Review by stefcameron — COMMENTED** [review]
And yes, this does break my extension, thank you for the head's up, @jim-docker , but thankfully the fix will be very easy when this ships. I just have to remember to check this PR to know what version it ships in.

File: src/extensions/registries/status-bar-registry.ts
Comment (jim-docker) [reply src/extensions/registries/status-bar-registry.ts:11]: It's consistent across all the registration types. Each of the XXXRegistration interfaces now has a `components` field, which map to interfaces specifying one or more components:

```
export interface PageComponents {
  Page: React.ComponentType<any>;
}
```

```
export interface AppPreferenceComponents {
  Hint: React.ComponentType<any>;
  Input: React.ComponentType<any>;
}
```

better to use `components` for all rather than `component` for some and `components` for others. Or do you suggest changing all of them to `component`?

Comment (stefcameron) [reply src/extensions/registries/status-bar-registry.ts:11]: Oh, I see. Consistency is key, for sure. What does `React.ComponentType<any>` represent, though. An array? If so, then I support the plural name. If not, why the "s"?

In any case, I'm probably fighting a losing battle since there is a lot of precedent if all other types are already using the plural form, and updating all of them would break a lot more stuff than just my extension.

**Comment by jakolehm** [comment]
> LGTM but @jakolehm should decide when this is merged

Code looks good but I'm not happy about the fact that we are willing to break the API here just before GA. #1606 contains minimal fix for the issue and that probably should be included in 4.0. This PR can be targeted for 4.1 if we come up a way to support multiple interfaces/versions (should be doable, see https://gist.github.com/jakolehm/7c79031856bab06e245d19b2eaa0ff38 for ideas).

**Comment by jim-docker** [comment]
This wouldn't have to breaking if it supported both:
```
export interface StatusBarRegistration {
  item?: React.ComponentType<any>;    // deprecated, use components field instead
  components?: {
    Item: React.ComponentType<any>;
  }
}
```
but then we'd have to tolerate releasing an api with a deprecated field...

**Comment by Nokel81** [comment]
That is probably what we would should do.

**Comment by stefcameron** [comment]
Jim's proposal is what I was going to propose.

**Comment by Nokel81** [comment]
I'll change my PR to that^^

**Review by roderickmarj — APPROVED** [review]

**Review by roderickmarj — APPROVED** [review]

**Review by roderickmarj — APPROVED** [review]

**Review by roderickmarj — APPROVED** [review]

**Review by roderickmarj — APPROVED** [review]

**Review by jakolehm — DISMISSED** [review]

File: docs/extensions/capabilities/common-capabilities.md
Comment (jim-docker) [inline docs/extensions/capabilities/common-capabilities.md]: mkdocs doesn't support tsx here, right?

File: docs/extensions/guides/renderer-extension.md
Comment (jim-docker) [inline docs/extensions/guides/renderer-extension.md:614]: @jakolehm, are the public docs for 4.0 pulled from master or a release branch? I think @pauljwil is making changes to docs (which would be good for 4.0.x and above) in master, but the change in this PR is strictly 4.1

If #1789 is meant for 4.1 only then there's no issue. If we want #1789 to be reflected in 4.0.x docs then we have to be careful here

File: docs/extensions/guides/renderer-extension.md
Comment (jim-docker) [inline docs/extensions/guides/renderer-extension.md:624]: The descriptive text below would have to change as well.

File: src/extensions/registries/status-bar-registry.ts
Comment (jim-docker) [inline src/extensions/registries/status-bar-registry.ts]: Does this _have_ to be exported? When it comes time to actually remove the `item` field, won't we just eliminate this interface and redefine `StatusBarRegistration` accordingly? If so it would be better if we didn't export something we'll eventually remove.

File: src/renderer/components/cluster-manager/bottom-bar.tsx
Comment (jim-docker) [inline src/renderer/components/cluster-manager/bottom-bar.tsx]: registrated? There are some obscure definitions for it. What about `renderRegisteredItem`?

File: docs/extensions/guides/renderer-extension.md
Comment (jakolehm) [reply docs/extensions/guides/renderer-extension.md:614]: Docs are generated from a release.

**Review by jim-docker — DISMISSED** [review]

**Review by jakolehm — DISMISSED** [review]

**Review by ixrock — DISMISSED** [review]

**Comment by Nokel81** [comment]
@jakolehm, @jim-docker, @ixrock could you PTL again, I needed to rebase to clear conflicts

File: docs/extensions/guides/renderer-extension.md
Comment (jim-docker) [inline docs/extensions/guides/renderer-extension.md]: Should be `Item` not `item`

**Review by jim-docker — DISMISSED** [review]

File: docs/extensions/capabilities/common-capabilities.md
Comment (jim-docker) [inline docs/extensions/capabilities/common-capabilities.md:43]: just curious, having a space is not OK? Seems to work fine

Comment (Nokel81) [reply docs/extensions/capabilities/common-capabilities.md:43]: It is fine, I was just trying to unify them.

**Review by jim-docker — APPROVED** [review]

**Review by jakolehm — APPROVED** [review]

**Pull request merged**
